sfputil/main.py

```diff
@@ -687,123 +693,247 @@ def eeprom(port, dump_dom, namespace):
+    """
+    Dump EEPROM for a single logical port in hex format.
+    Args:
+        logical_port_name: logical port name
+        page: page to be dumped
+
+    Returns:
+        tuple(0, dump string) if success else tuple(error_code, error_message)
+    """
+    if platform_sfputil.is_logical_port(logical_port_name) == 0:
+        print_all_valid_port_values()
+        return ERROR_INVALID_PORT, f'Error: invalid port {logical_port_name}'
 
     if is_port_type_rj45(logical_port_name):
-        click.echo("{}: SFP EEPROM Hexdump is not applicable for RJ45 port".format(port))
-        sys.exit(ERROR_INVALID_PORT)
+        return ERROR_INVALID_PORT, f'{logical_port_name}: SFP EEPROM Hexdump is not applicable for RJ45 port'
 
+    physical_port = logical_port_to_physical_port_index(logical_port_name)
     try:
-        presence = platform_chassis.get_sfp(physical_port).get_presence()
+        sfp = platform_chassis.get_sfp(physical_port)
+        presence = sfp.get_presence()
     except NotImplementedError:
-        click.echo("Sfp.get_presence() is currently not implemented for this platform")
-        sys.exit(ERROR_NOT_IMPLEMENTED)
+        return ERROR_NOT_IMPLEMENTED, 'Sfp.get_presence() is currently not implemented for this platform'
 
     if not presence:
-        click.echo("SFP EEPROM not detected")
-        sys.exit(ERROR_NOT_IMPLEMENTED)
-    else:
-        try:
-            id = platform_chassis.get_sfp(physical_port).read_eeprom(0, 1)
-            if id is None:
-                click.echo("Error: Failed to read EEPROM for offset 0!")
-                sys.exit(ERROR_NOT_IMPLEMENTED)
-        except NotImplementedError:
-            click.echo("Sfp.read_eeprom() is currently not implemented for this platform")
-            sys.exit(ERROR_NOT_IMPLEMENTED)
+        return ERROR_NOT_IMPLEMENTED, 'SFP EEPROM not detected'
 
-        if id[0] == 0x3:
-            output = eeprom_hexdump_sff8472(port, physical_port, page)
+    try:
+        api = sfp.get_xcvr_api()
+        if not api:
+            return ERROR_NOT_IMPLEMENTED, 'Error: Failed to read EEPROM for offset 0!'
+
+        from sonic_platform_base.sonic_xcvr.api.public import sff8636, sff8436, cmis, sff8472
+        from sonic_platform_base.sonic_xcvr.fields import consts
+        if isinstance(api, cmis.CmisApi):
+            if page is None: # print all possible pages
+                if api.is_flat_memory():
+                    pages = [0]
+                else:
+                    pages = copy.deepcopy(CMIS_MODULE_PAGES)
+                    if api.is_coherent_module():
+                        pages.extend(CMIS_COHERENT_MODULE_PAGES)
+                    cdb_support = api.xcvr_eeprom.read(consts.CDB_SUPPORT)
+                    if cdb_support != 0:
+                        pages.append(0x9f)
+            else:
+                pages = [0]
+                if page not in pages:
+                    pages.append(page)
```

Comment:
@Junchao-Mellanox why display upper-page 0 always even if user wants to see upper-page >] 1?

Reply:
For backward compatible, previous implementation always print page 0. As you can see, I did not change the previous unit test cases, and all passed.